zebra-chain/src/addresses/sprout.rs

```diff
@@ -0,0 +1,135 @@
+//! Sprout Shielded Payment Address types.
+
+use std::{fmt, io};
+
+use bs58;
+
+#[cfg(test)]
+use proptest::{arbitrary::Arbitrary, array, prelude::*};
+
+use crate::{
+    keys::sprout,
+    serialization::{ReadZcashExt, SerializationError, ZcashDeserialize, ZcashSerialize},
+    Network,
+};
+
+/// Magic numbers used to identify what networks Sprout Shielded
+/// Addresses are associated with.
+mod magics {
+    pub const MAINNET: [u8; 2] = [0x16, 0x9A];
+    pub const TESTNET: [u8; 2] = [0x16, 0xB6];
+}
+
+/// Sprout Shielded Payment Addresses
+///
+/// In Bitcoin a single byte is used for the version field identifying
+/// the address type. In Zcash two bytes are used. For addresses on
+/// the production network, this and the encoded length cause the first
+/// two characters of the Base58Check encoding to be xed as “t3” for
+/// P2SH addresses, and as “t1” for P2PKH addresses. (This does not
+/// imply that a transparent Zcash address can be parsed identically
+/// to a Bitcoin address just by removing the “t”.)
+///
+/// https://zips.z.cash/protocol/protocol.pdf#transparentaddrencoding
+#[derive(Copy, Clone)]
+pub struct SproutShieldedAddress {
+    network: Network,
+    paying_key: sprout::PayingKey,
+    transmission_key: sprout::TransmissionKey,
+}
+
+impl fmt::Debug for SproutShieldedAddress {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut bytes = io::Cursor::new(Vec::new());
+        let _ = self.zcash_serialize(&mut bytes);
+
+        f.debug_tuple("SproutShieldedAddress")
+            .field(&bs58::encode(bytes.get_ref()).with_check().into_string())
+            .finish()
```

Comment:
Isn't a representation of the parsed fields more useful for debugging? (They might have been parsed incorrectly, or this code might reserialize them incorrectly!)

Reply:
This follows on from the impl for transparent addresses where the debug is showing literally the t/z-addr, for now.

Reply:
It will probably behoove us to have `Debug` do what you suggest and `Display` do what's happening here.

Reply:
Implementing `FromStr` so that `FromStr` roundtrips with `Display` would be good too.